geopandas/tools/clip.py

```diff
@@ -0,0 +1,246 @@
+    sidx = list(spatial_index.intersection(bbox))
+    gdf_sub = gdf.iloc[sidx]
+
+    # Clip the data with the polygon
+    if isinstance(gdf_sub, GeoDataFrame):
+        clipped = gdf_sub.copy()
+        clipped["geometry"] = gdf_sub.intersection(poly)
+
+        # Return the clipped layer with no null geometry values or empty geometries
+        return clipped[~clipped.geometry.is_empty & clipped.geometry.notnull()]
+    else:
+        # GeoSeries
+        clipped = gdf_sub.intersection(poly)
+        return clipped[~clipped.is_empty & clipped.notnull()]
+
+
+def clip(gdf, clip_obj, keep_geom_type=False):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS).
+    The `gdf` will be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj, data from `gdf` will be
+    clipped to the total boundary of all polygons in clip_obj.
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame or GeoSeries
+        Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+        Polygon vector layer used to clip `gdf`.
+        The clip_obj's geometry is dissolved into one geometric feature
+        and intersected with `gdf`.
+    keep_geom_type : boolean
+        If True, return only geometries of original type in case of intersection
+        resulting in multiple geometry types or GeometryCollections.
+        If False, return all resulting geometries (potentially mixed-types).
+
+    Returns
+    -------
+    GeoDataFrame or GeoSeries
+         Vector data (points, lines, polygons) from `gdf` clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global cities) with a polygon (the South American continent):
+
+    >>> import geopandas
+    >>> path =
+    >>> world = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_lowres'))
+    >>> south_america = world[world['continent'] == "South America"]
+    >>> capitals = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_cities'))
+    >>> capitals.shape
+    (202, 2)
+    >>> sa_capitals = geopandas.clip(capitals, south_america)
+    >>> sa_capitals.shape
+    (12, 2)
+    """
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(clip_obj, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'clip_obj' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(clip_obj, (GeoDataFrame, GeoSeries)):
+        xmin, ymin, xmax, ymax = clip_obj.total_bounds
+    else:
+        xmin, ymin, xmax, ymax = clip_obj.bounds
+    if gdf.cx[xmin:xmax, ymin:ymax].empty:
```

Comment:
Since both `_clip_line_poly` and `_clip_points` evaluate their input against the spatial index, I think it makes sense to (1) compute that a single time before delegating to the `_clip` and (2) use `index_hits.any()` as the check on whether or not to stop early? 

This makes the masks to `_clip_line_poly` and `_clip_points` become `poly_idx & hits_index` (or, `points_idx & hits_index`) That  would do the expensive thing once (evaluate the sindex) and uses the result everywhere it's needed.

Reply:
Thanks, that sounds as a good idea!

Reply:
@ljwolf but in case of non-overlapping gdfs, you generate sindex to figure out that there is nothing to clip, which is not needed. I think that @jorisvandenbossche's suggestion of bbox overlap is better in such a case. I'll make a commit.